File: engine/mutate/groups.py
"""The client's own grouping inside a tax category (§13.1).

A group is a reporting dimension and nothing else: it never carries a rate, a
repair limit or a figure of any kind. The moment it did, it would be a
substitute category and the aggregation by art. 114/115 would stop being
universal -- which is exactly what the fixed category enum in §4 protects.

So everything here is bookkeeping about names: create one, rename it, delete
it, put cards into it. No calculation reads any of it."""

from __future__ import annotations

from pathlib import Path

from ..storage import DataError

from .core import rows_of, save_rows, transaction

GROUP_ID_FMT = "QR-{:03d}"


def next_group_id(rows: list[dict[str, str]]) -> str:
    n = 0
    for r in rows:
        gid = r.get("group_id", "")
        if gid.startswith("QR-") and gid[3:].isdigit():
            n = max(n, int(gid[3:]))
    return GROUP_ID_FMT.format(n + 1)


def find_group(rows: list[dict[str, str]], name: str) -> dict[str, str] | None:
    """A group by name, case- and space-insensitively.

    The whole point of keeping a dictionary is that «Serverlər» typed twice is
    one group, so matching has to be as forgiving as the typing is (§2.1: a
    silently split report is worse than none).
    """
    key = " ".join(str(name or "").split()).casefold()
    return next((r for r in rows
                 if " ".join(r["name"].split()).casefold() == key), None)


def group_ref(root: Path, slug: str, value) -> str:
    """Validate a group_id coming in from a form. Empty means "no group".

    Checked at the door as well as at read time (§8): the recompute after a
    write would catch a dangling reference anyway, but only by rolling the
    whole action back with a message about the store, not about the field the
    user just filled in.
    """
    gid = str(value or "").strip()
    if not gid:
        return ""
    if not any(r["group_id"] == gid for r in rows_of(root, slug, "groups.tsv")):
        raise DataError(f"növ tapılmadı: {gid}")
    return gid


def _clean_name(value) -> str:
    name = " ".join(str(value or "").split())
    if not name:
        raise DataError("Növün adı boş ola bilməz")
    return name


def create_group(root: Path, slug: str, p: dict) -> str:
    with transaction(root, slug, "group.create") as tx:
        groups = rows_of(root, slug, "groups.tsv")
        name = _clean_name(p.get("name"))
        hit = find_group(groups, name)
        if hit:
            raise DataError(f"«{hit['name']}» növü artıq mövcuddur")
        gid = next_group_id(groups)
        groups.append({"group_id": gid, "name": name,
                       "note": str(p.get("note", "")).strip()})
        save_rows(root, slug, "groups.tsv", groups)
        tx.log("", "group", "", f"{gid} {name}")
    return gid


def update_group(root: Path, slug: str, p: dict) -> str:
    """Rename a group. One edit here instead of one per card -- the reason the
    card stores `group_id` and not the name (§4, the inv_no reasoning)."""
    gid = str(p.get("group_id", "")).strip()
    with transaction(root, slug, "group.update") as tx:
        groups = rows_of(root, slug, "groups.tsv")
        row = next((r for r in groups if r["group_id"] == gid), None)
        if row is None:
            raise DataError(f"növ tapılmadı: {gid}")
        name = _clean_name(p.get("name"))
        hit = find_group(groups, name)
        if hit and hit["group_id"] != gid:
            raise DataError(f"«{hit['name']}» növü artıq mövcuddur")
        note = str(p.get("note", "")).strip()
        for field, value in (("name", name), ("note", note)):
            if row[field] != value:
                tx.log("", f"group.{field}", row[field], value)
                row[field] = value
        save_rows(root, slug, "groups.tsv", groups)
    return gid


def delete_group(root: Path, slug: str, p: dict) -> str:
    """Refused while cards still point at it.

    Clearing forty cards as a side effect of one click is data loss that looks
    like tidying up: the group is gone, the cards silently lose their only
    non-tax classification, and nothing on screen says so. Emptying it first
    is one action away (assign them elsewhere), and then this is safe.
    """
    gid = str(p.get("group_id", "")).strip()
    with transaction(root, slug, "group.delete") as tx:
        groups = rows_of(root, slug, "groups.tsv")
        row = next((r for r in groups if r["group_id"] == gid), None)
        if row is None:
            raise DataError(f"növ tapılmadı: {gid}")
        assets = rows_of(root, slug, "assets.tsv")
        used = [r for r in assets if r.get("group_id") == gid]
        if used:
            raise DataError(
                f"«{row['name']}» növündə {len(used)} ƏV var — əvvəlcə onları "
                f"başqa növə keçirin və ya növü boşaldın"
            )
        tx.log("", "group", row["name"], "silindi")
        save_rows(root, slug, "groups.tsv",
                  [r for r in groups if r["group_id"] != gid])
    return gid


def assign_group(root: Path, slug: str, p: dict) -> str:
    """Put a batch of cards into a group (or take them out of one).

    A list rather than one card at a time for the reason §5.3-bis gives for
    set_writeoff: every write backs the folder up and recomputes every open
    year (§8.1), so forty separate calls would mean forty backups for one act
    of sorting. An empty `group_id` clears the field.

    A closed year is NOT a barrier here, unlike cost or date: the group
    reaches no figure, so putting a 2024 asset into «Serverlər» cannot change
    a filed return. Closing seals the return, not the card (§6.2).
    """
    ids = p.get("asset_ids") or ([p["asset_id"]] if p.get("asset_id") else [])
    gid = str(p.get("group_id", "")).strip()
    if not ids:
        raise DataError("ƏV seçilməyib")
    with transaction(root, slug, "group.assign") as tx:
        groups = rows_of(root, slug, "groups.tsv")
        if gid and not any(r["group_id"] == gid for r in groups):
            raise DataError(f"növ tapılmadı: {gid}")
        name = next((r["name"] for r in groups if r["group_id"] == gid), "")
        assets = rows_of(root, slug, "assets.tsv")
        by_id = {r["asset_id"]: r for r in assets}
        touched = 0
        for aid in ids:
            row = by_id.get(str(aid))
            if row is None:
                raise DataError(f"ƏV tapılmadı: {aid}")
            if row.get("group_id", "") == gid:
                continue
            # One line per card even though the act was one, same as a batch
            # purchase: grouping is for the person, not a licence to record
            # less of what happened (§5.3-bis).
            tx.log(row["asset_id"], "group_id", row.get("group_id", ""), gid)
            row["group_id"] = gid
            touched += 1
        save_rows(root, slug, "assets.tsv", assets)
    return f"{touched} ƏV → {name or '— növsüz —'}"

File: web/static/groups.js
/* ---- the client's own grouping, beside the tax category (§13.1) -------

   The tax category belongs to the law: seven codes, fixed, and every figure
   in the return is aggregated by them. «Növ» belongs to the client:
   notebooks, servers, printers -- all of them `yt` at one and the same rate,
   and the office still needs them apart.

   The line between the two is the whole design, and it runs here: a group
   never carries a rate, a repair limit or any other number. It is a way of
   CUTTING the report, not a tax fact. The moment it acquired a rate it would
   be a substitute category and the aggregation by art. 114/115 would stop
   being universal -- exactly what the fixed enum in §4 protects.

   Kept as a dictionary (groups.tsv) rather than free text on the card, for
   the same reason `category` is an enum: "Serverlər" and "Serverler" typed on
   two different days would silently become two groups, and a report split in
   half is worse than no report (§2.1). Renaming is then one edit here, not
   forty in the cards. ------------------------------------------------- */

const NOGROUP = '__none__';                    // the cards in no group at all

const GROUPS = () => (REPORT && REPORT.groups) || [];
const groupName = id => (GROUPS().find(g => g.group_id === id) || {}).name || '';

/* Every card of the year, flat: a group cuts ACROSS categories, so this is
   the list to count over and to choose from. */
const allCards = () => (REPORT ? REPORT.categories.flatMap(c => c.cards) : []);
const groupCount = id => allCards().filter(
  c => id === NOGROUP ? !c.group_id : c.group_id === id).length;

/* ---- the dictionary ---- */
function formGroups(){
  const list = GROUPS();
  const rows = list.map(g => `<tr>
      <td><strong>${esc(g.name)}</strong>
        ${g.note ? `<div class="h">${esc(g.note)}</div>` : ''}</td>
      <td class="num">${groupCount(g.group_id)}</td>
      <td class="acts" style="justify-content:flex-end">
        <button type="button" class="tagbtn"
          onclick="formAssign('${g.group_id}')">ƏV seç</button>
        <button type="button" class="tagbtn"
          onclick="formGroupEdit('${g.group_id}')">adı</button>
        <button type="button" class="tagbtn"
          onclick="formGroupDelete('${g.group_id}')">sil</button></td>
    </tr>`).join('');
  const none = groupCount(NOGROUP);

  openModal('Növlər — müştərinin öz bölgüsü', `
    <div class="note">Növ <strong>heç bir rəqəmə təsir etmir</strong>: dərəcə,
      m.115 limiti və bəyannamə sətirləri həmişə vergi kateqoriyası üzrə
      hesablanır. Növ hesabatı kəsmək üçündür — süzgəc, aralıq yekunlar və
      Excel-də ayrıca sütun.</div>
    ${list.length ? `<div class="scroll" style="max-height:38vh"><table>
      <thead><tr><th>Ad</th><th class="num">ƏV</th><th></th></tr></thead>
      <tbody>${rows}${none ? `<tr><td class="h">— növsüz —</td>
        <td class="num">${none}</td>
        <td class="acts" style="justify-content:flex-end"><button type="button"
          class="tagbtn" onclick="formAssign('${NOGROUP}')">növü sil</button></td>
        </tr>` : ''}</tbody></table></div>`
      : `<p class="h">Hələ növ yaradılmayıb. Məsələn: «Noutbuklar»,
         «Server avadanlığı», «Printerlər».</p>`}
    <hr style="border:0;border-top:1px solid var(--line);margin:14px 0">
    ${fld('name','Yeni növ',{placeholder:'məs. Server avadanlığı'})}`,
    async d => {
      if (!(d.name || '').trim()) throw new Error('Növün adı boş ola bilməz');
      return post('group.create', {name: d.name});
    }, '+ Növ əlavə et');
}

function formGroupEdit(gid){
  const g = GROUPS().find(x => x.group_id === gid);
  openModal(`Növ — ${g.name}`,
    `<input type="hidden" name="group_id" value="${gid}">
     ${fld('name','Ad',{value:g.name,req:true})}
     ${fld('note','Qeyd',{value:g.note})}
     <div class="h">Ad kartlarda saxlanmır, ona görə dəyişiklik bütün
       ƏV-lərdə dərhal görünür — bu, adın deyil, nömrənin kartda
       saxlanmasının səbəbidir.</div>`,
    d => post('group.update', d), 'Yadda saxla');
}

function formGroupDelete(gid){
  const g = GROUPS().find(x => x.group_id === gid);
  const n = groupCount(gid);
  openModal(`«${g.name}» növünü sil`,
    `<input type="hidden" name="group_id" value="${gid}">
     <div class="note">${n
        ? `Bu növdə <strong>${n} ƏV</strong> var və silinmə qəbul edilməyəcək:
           bir kliklə ${n} kartın bölgüsü səssizcə itərdi. Əvvəlcə «ƏV seç»
           ilə onları başqa növə keçirin.`
        : 'Növ boşdur. Silinmə ƏV kartlarına toxunmur.'}</div>`,
    d => post('group.delete', d), 'Sil');
}

/* ---- putting cards INTO a group ----
   One action for the whole selection, not one per card: every write backs the
   folder up and recomputes each open year (§8.1), so forty separate calls
   would mean forty backups for one act of sorting. Same reasoning as the
   500/5% list (§5.3-bis). */
function formAssign(gid){
  const target = gid === NOGROUP ? '' : gid;
  const rows = allCards().map(c => `<tr>
      <td><input type="checkbox" data-aid="${c.asset_id}" style="width:auto"
            ${c.group_id === target ? 'disabled checked' : ''}></td>
      <td>${esc(c.inv_no)}</td><td>${esc(c.name)}</td>
      <td>${esc(c.category)}</td>
      <td class="h">${esc(c.group || '—')}</td></tr>`).join('');

  openModal(target ? `«${groupName(target)}» növünə keçir` : 'Növsüz et', `
    <input type="hidden" name="group_id" value="${target}">
    <div class="fld"><label>Süzgəc</label>
      <input id="assignq" placeholder="ad, inv.№, kateqoriya, indiki növ…"
        oninput="assignFilter(this.value)"></div>
    <div class="acts" style="margin:0 0 8px">
      <button type="button" class="tagbtn" onclick="assignAll(true)"
        >görünənləri seç</button>
      <button type="button" class="tagbtn" onclick="assignAll(false)"
        >seçimi ləğv et</button>
      <span class="h" id="assigncount"></span></div>
    <div class="scroll" style="max-height:38vh" id="assignbox"><table>
      <thead><tr><th></th><th>İnv.№</th><th>Adı</th><th>Kat.</th>
        <th>İndiki növ</th></tr></thead><tbody>${rows}</tbody></table></div>`,
    d => {
      const ids = assignPicked();
      if (!ids.length) throw new Error('ƏV seçilməyib');
      return post('group.assign', {group_id: d.group_id, asset_ids: ids});
    }, 'Keçir');
  document.getElementById('assignbox').onchange = assignCount;
  assignCount();
}

const assignPicked = () =>
  [...document.querySelectorAll('#assignbox input:checked')]
    .filter(el => !el.disabled).map(el => el.dataset.aid);

function assignFilter(q){
  const s = q.trim().toLowerCase();
  for (const tr of document.querySelectorAll('#assignbox tbody tr'))
    tr.style.display = !s || tr.textContent.toLowerCase().includes(s) ? '' : 'none';
}

/* "Select all" means the rows in front of you, not the rows the filter is
   hiding -- selecting what cannot be seen is how a bulk action surprises
   someone. */
function assignAll(on){
  for (const tr of document.querySelectorAll('#assignbox tbody tr')){
    if (tr.style.display === 'none') continue;
    const box = tr.querySelector('input');
    if (!box.disabled) box.checked = on;
  }
  assignCount();
}

function assignCount(){
  const box = document.getElementById('assigncount');
  if (box) box.textContent = assignPicked().length
    ? `${assignPicked().length} ƏV seçilib` : '';
}

/* ---- the field on the card form ----
   A dropdown of the dictionary plus one explicit "new" entry. A free text box
   is precisely what turns one group into two. */
function groupField(card){
  const cur = (card && card.group_id) || '';
  const opts = ['<option value="">— növsüz —</option>']
    .concat(GROUPS().map(g => `<option value="${g.group_id}"
      ${cur === g.group_id ? 'selected' : ''}>${esc(g.name)}</option>`))
    .concat(['<option value="__new__">+ yeni növ…</option>']).join('');
  return `<div class="fld"><label>Növ <span class="h">— müştərinin öz
      bölgüsü, heç bir rəqəmə təsir etmir</span></label>
    <select name="group_id" onchange="groupPick(this)">${opts}</select>
    <input name="group_new" placeholder="yeni növün adı"
      style="display:none;margin-top:6px"></div>`;
}

function groupPick(sel){
  const box = sel.parentNode.querySelector('input[name=group_new]');
  box.style.display = sel.value === '__new__' ? '' : 'none';
  if (sel.value === '__new__') box.focus();
}

/* Creating the group is its own act, so it is its own write -- and it goes
   first, because the card needs the id. Two backups instead of one, and only
   when a new name is actually introduced. */
async function groupResolve(d){
  if (d.group_id === '__new__'){
    const name = (d.group_new || '').trim();
    if (!name) throw new Error('Yeni növün adı boşdur');
    d.group_id = (await post('group.create', {name})).result;
  }
  delete d.group_new;
  return d;
}
